tensorflow_datasets/image/flowers.py

```diff
@@ -46,49 +51,50 @@ def _info(self):
     return tfds.core.DatasetInfo(
         builder=self,
         description="A large set of images of flowers",
-        features=tfds.features.FeaturesDict({
-            "image": tfds.features.Image(),
-            "label": tfds.features.ClassLabel(
-                names=["dandelion", "daisy", "tulips", "sunflowers", "roses"]),
-        }),
+        features=tfds.features.FeaturesDict(
+            {
+                "image": tfds.features.Image(),
+                "label": tfds.features.ClassLabel(
+                    names=["dandelion",
+                           "daisy",
+                           "tulips",
+                           "sunflowers",
+                           "roses",]
+                ),
+            }
+        ),
         supervised_keys=("image", "label"),
         homepage="https://www.tensorflow.org/tutorials/load_data/images",
         citation=_CITATION
         )
 
   def _split_generators(self, dl_manager):
-    path = dl_manager.download_and_extract(_URL)
-
+    path = dl_manager.download(_URL)
+    if tf.io.gfile.isdir(path):
+      path = os.path.join(path, tf.io.gfile.listdir(path)[0])
     # There is no predefined train/val/test split for this dataset.
     return [
         tfds.core.SplitGenerator(
             name=tfds.Split.TRAIN,
-            gen_kwargs={
-                "images_dir_path": path
-            }),
+            num_shards=20,
+            gen_kwargs={"archive": dl_manager.iter_archive(path)},
+        )
     ]
 
-  def _generate_examples(self, images_dir_path):
+  def _generate_examples(self, archive):
     """Generate flower images and labels given the image directory path.
 
-    Args:
-      images_dir_path: path to the directory where the images are stored.
+  Args:
+    archive: Complete path where the image is stored.
 
-    Yields:
-      The image path and its corresponding label.
+  Yields:
+    The image path and its corresponding label.
     """
-    parent_dir = tf.io.gfile.listdir(images_dir_path)[0]
-    walk_dir = os.path.join(images_dir_path, parent_dir)
-    dirs = tf.io.gfile.listdir(walk_dir)
-
-    for d in dirs:
-      if tf.io.gfile.isdir(os.path.join(walk_dir, d)):
-        for full_path, _, fname in tf.io.gfile.walk(os.path.join(walk_dir, d)):
-          for image_file in fname:
-            if image_file.endswith(".jpg"):
-              image_path = os.path.join(full_path, image_file)
-              record = {
-                  "image": image_path,
-                  "label": d.lower(),
-              }
-              yield "%s/%s" % (d, image_file), record
+    for fname, fobj in archive:
+      res = _NAME_RE.match(fname)
```

Comment:
I think regex are less readable, more error prone and overkill for this. I prefer the `image_file.endswith(".jpg")` check and the label could be extracted from the dir name (`fname.split('/')`[-2])